### PR TITLE
closes 3305 adds customize support

### LINKED ISSUE #3305: vmware_guest - set static IP for virtual machine
##### ISSUE TYPE

- Feature Idea
##### COMPONENT NAME

vmware_guest
##### SUMMARY

Would like to be able to set the IP address of a VM when cloning from a template.  Is this already possible?  How are other people solving this??

### PR DESCRIPTION
**Pull request by dav1x**
##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

vmware_guest

##### ANSIBLE VERSION

```
ansible 2.2.0.0

```

##### SUMMARY

Fixes #3305

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @dav1x for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

File: cloud/vmware/vmware_guest.py [hunk 1]
```diff
@@ -89,6 +89,26 @@
         description:
             - The esxi hostname where the VM will run.
         required: True
+   customize:
```
Comment (jctanner) [inline cloud/vmware/vmware_guest.py:92]: Need to have a version_added field for these new parameters.

File: cloud/vmware/vmware_guest.py [hunk 2]
```diff
@@ -672,6 +715,109 @@ def deploy_template(self, poweron=False, wait_for_ip=False):
...
+                                if hasattr(device, 'addressType'):
+                                    nic = vim.vm.device.VirtualDeviceSpec()
+                                    nic.operation = \
+                                        vim.vm.device.VirtualDeviceSpec.Operation.remove
+                                    nic.device = device
+                                    devices.append(nic)
+                except:
+                        pass
+
+                    # for key, ip in enumerate(ip_settings):
```
Comment (gundalow) [inline cloud/vmware/vmware_guest.py]: Please delete commented out code

Comment (dav1x) [reply cloud/vmware/vmware_guest.py]: Done!

File: cloud/vmware/vmware_guest.py (hunk 1 above)
Comment (dav1x) [reply cloud/vmware/vmware_guest.py:92]: done!

**Comment by aperigault** [comment]
Hello,

I think it could be usefull to add nics without enable customization. Indeed, OS like debian do not support OS customization but I still need to add nics.

**Comment by jctanner** [comment]
@aperigault I agree. Is that something you can send a patch for or can you collaborate with @dav1x ?

**Comment by dav1x** [comment]
@aperigault @jctanner The first half of the customization spec adds the NIC first. It wouldn't be too difficult to call that outside of the Customize: True operation.

**Comment by aperigault** [comment]
@jctanner @dav1x I do a PR to do this + fix bugs. Tell me what you thinking about it
#3481

**Comment by angystardust** [comment]
Hi @dav1x, unfortunately it doesn't seem to work:

```
[ansible@bastion vmware]$ ansible-playbook --ask-vault-pass create-vm.yml --extra-vars 'name=test01'
Vault password: 

PLAY [VM from template] ********************************************************

TASK [Check for vSphere access parameters] *************************************
skipping: [localhost]

TASK [Create VM from template] *************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: KeyError: 'network'
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_og2HvC/ansible_module_vmware_guest.py\", line 1152, in <module>\n    main()\n  File \"/tmp/ansible_og2HvC/ansible_module_vmware_guest.py\", line 1130, in main\n    wait_for_ip=module.params['wait_for_ip_address']\n  File \"/tmp/ansible_og2HvC/ansible_module_vmware_guest.py\", line 726, in deploy_template\n    if self.params['ips'] and self.params['network']:\nKeyError: 'network'\n", "module_stdout": "", "msg": "MODULE FAILURE"}
    to retry, use: --limit @/srv/ansible/infra/playbooks/vmware/create-vm.retry

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   

```

This is my playbook:

```
---
#
# Create a new VM from a template

- name: VM from template
  hosts: localhost
  gather_facts: false
  connection: local

  vars:
    esxhost: esx1.example.com
    datastore: "netapp01"
    network: "Dev"
    vmtemplate: template-CentOS7.2x64
    my_datacenter: MAIN

  vars_files:
    - /srv/ansible/private/vmware.yml

  tasks:

    - name: Check for vSphere access parameters
      fail: msg="Must set vcenter_user and vcenter_pass in a Vault"
      when: (vcenter_user is not defined) or (vcenter_pass is not defined)

    - name: Create VM from template
      vmware_guest:
        validate_certs: False
        hostname: "{{ vcenter_hostname }}"
        username: "{{ vcenter_user }}"
        password: "{{ vcenter_pass }}"
        esxi_hostname: "{{ esxhost }}"
        datacenter: "{{ my_datacenter }}"
        name: "{{ name }}"
        template: "{{ vmtemplate }}"
        customize: True
        domain: "example.com"
        dns_servers: ['192.168.1.1','192.168.1.2']
        ips: "192.168.1.100"
        networks:
          '192.168.1.0/24':
            network: 'VM Network'
            gateway: '192.168.1.1' 
        nic:
          - type: vmxnet3
            network: "{{ network }}"
        hardware:
          memory_mb: "{{ vm_memory | default(1024) }}"
        wait_for_ip_address: False
        state: present
      register: newvm

    - name: IP address info
      debug:
        msg: "{{ newvm.instance.ipv4 }} {{ name }}"

```

**Comment by jbilbro** [comment]
@angystardust I had that same error before @dav1x made a fix for it yesterday.  I'm pretty sure it got merged - have you done a git pull and updated your local copy?

**Comment by angystardust** [comment]
Hi @jbilbro i've just noticed a related commit that was merged yesterday: https://github.com/ansible/ansible-modules-extras/pull/3477
I will try to make a git pull an test again my playbook
Thanks a lot for your tip ;)

**Comment by AlekseyKrylov** [comment]
Hi all!

Is it possible to add support of customization template? 

it's necessary when we unable set network directly, for example by security reason (vmware infrastucture managed by Virtualization Admins only; client has access to start/stop vm's and clone from template with predefined clone customization template)

**Comment by DanyC97** [comment]
any idea on which release will this make PR make it ? 2.3 maybe ?

**Pull request merged**